dali/kernels/shape.h

```diff
@@ -0,0 +1,609 @@
+  assert(0 <= count && count <= size() &&
+         "Number of dimensions in subshape must be between 0 and size()");
+  TensorShape<DynamicDimensions> result;
+  result.resize(count);
+  for (int i = 0; i < count; i++) {
+    result[i] = (*this)[i];
+  }
+  return result;
+}
+
+template <typename Container, int ndim>
+TensorShape<DynamicDimensions> TensorShapeBase<Container, ndim>::last(int count) {
+  assert(0 <= count && count <= size() &&
+         "Number of dimensions in subshape must be between 0 and size()");
+  TensorShape<DynamicDimensions> result;
+  result.resize(count);
+  int start_offset = size() - count;
+  for (int i = 0; i < count; i++) {
+    result[i] = (*this)[start_offset + i];
+  }
+  return result;
+}
+
+template <int left_ndim, int right_ndim>
+bool operator==(const TensorShape<left_ndim> &left, const TensorShape<right_ndim> &right) {
+  if (left.size() != right.size()) {
+    return false;
+  }
+  int size = left.size();
+  for (int i = 0; i < size; i++) {
+    if (left[i] != right[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <int left_ndim, int right_ndim>
+bool operator!=(const TensorShape<left_ndim> &left, const TensorShape<right_ndim> &right) {
+  return !(left == right);
+}
+
+constexpr int shape_cat_ndim(int left_ndim, int right_ndim) {
+  return (left_ndim == DynamicDimensions || right_ndim == DynamicDimensions)
+             ? DynamicDimensions
+             : left_ndim + right_ndim;
+}
+
+template <int left_ndim, int right_ndim>
+TensorShape<shape_cat_ndim(left_ndim, right_ndim)> shape_cat(const TensorShape<left_ndim> &left,
+                                                             const TensorShape<right_ndim> &right) {
+  TensorShape<shape_cat_ndim(left_ndim, right_ndim)> result;
+  int total_size = left.size() + right.size();
+  result.resize(total_size);
+  for (int i = 0; i < left.size(); i++) {
+    result[i] = left[i];
+  }
+  for (int i = 0; i < right.size(); i++) {
+    result[left.size() + i] = right[i];
+  }
+  return result;
+}
+
+template <int sample_ndim>
+typename std::enable_if<sample_ndim != DynamicDimensions, std::vector<int64_t>>::type
+flatten_shapes(const std::vector<TensorShape<sample_ndim>> &shapes) {
+  std::vector<int64_t> result;
+  result.resize(sample_ndim * shapes.size());
+  for (int sample = 0; sample < shapes.size(); sample++) {
+    for (int axis = 0; axis < sample_ndim; axis++) {
+      result[sample * sample_ndim + axis] = shapes[sample][axis];
+    }
+  }
+  return result;
+}
+
+template <typename T>
+typename std::enable_if<std::is_same<T, TensorShape<DynamicDimensions>>::value ||
+                            std::is_same<T, std::vector<int64_t>>::value,
+                        int>::type
+get_dim_from_uniform(const std::vector<T> &shapes) {
+  if (shapes.empty()) {
+    return 0;
+  }
+  return shapes[0].size();
+}
+
+template <typename T>
+typename std::enable_if<std::is_same<T, TensorShape<DynamicDimensions>>::value ||
```

Comment:
Why `is_same`? We're copying anyway, so `is_integral` would be more appropriate.